hare/algorithm.go

```
@@ -0,0 +1,230 @@
+
+	// TODO: decide how we store messages maybe we don't need raw msgs
+	proc.knowledge = append(proc.knowledge, m)
+}
+
+func (proc *ConsensusProcess) nextRound() {
+	log.Info("End of round: %d", proc.k)
+
+	// send message if available
+	if proc.roundMsg != nil {
+		proc.network.Broadcast(ProtoName, proc.roundMsg)
+		proc.roundMsg = nil
+	}
+
+	// advance to next round
+	proc.k++
+
+	// reset knowledge
+	proc.knowledge = make([]*pb.HareMessage, 0, InitialKnowledgeSize)
+}
+
+// TODO: put state and msg in Context struct?
+func (proc *ConsensusProcess) processMessageByRound(msg *pb.HareMessage) {
+	// TODO: do we have cases in which we don't need to do further processing? (final decisions)
+	switch proc.k % 4 { // switch round
+	case 0: // end of round 0
+		proc.processMsgRound0(msg)
+	case 1: // end of round 1
+		proc.processMsgRound1(msg)
+	case 2: // end of round 2
+		proc.processMsgRound2(msg)
+	case 3: // end of round 3
+		proc.processMsgRound3(msg)
+	}
+}
+
+// TODO: maybe change it to SetStatusMessage and just set proc.roundMsg?
+func (proc *ConsensusProcess) buildStatusMessage() ([]byte, error) {
+	x := NewInnerBuilder().SetType(Status).SetLayer(proc.layerId).SetIteration(proc.k).SetKi(proc.ki).SetBlocks(proc.s).Build()
+
+	buff, err := proto.Marshal(x)
+	if err != nil {
+		return nil, errors.New("error marshaling inner message")
+	}
+
+	sig := proc.signing.Sign(buff)
+
+	outer := NewOuterBuilder()
+	m := outer.SetInnerMessage(x).SetPubKey(proc.pubKey).SetInnerSignature(sig).Build()
+
+	buff, err = proto.Marshal(m)
+	if err != nil {
+		return nil, errors.New("error marshaling message")
+	}
+
+	return buff, nil
+}
+
+func (proc *ConsensusProcess) updateState(state State) {
+	proc.stateMutex.Lock()
+	proc.State = state
+	proc.stateMutex.Unlock()
+}
+
+func roleFromIteration(k uint32) byte {
+	if k%4 == 1 { // only round 1 is leader
+		return Leader
+	}
+
+	return Active
+}
+
+func (proc *ConsensusProcess) processMsgRound0(msg *pb.HareMessage) {
```

you'd probable want to return an error as well

---

This is merely a structure to show the flow. When implemented we will see what it requires.